### PR TITLE
solveset and hyperbolic/trig modification

### LINKED ISSUE #9531: solveset(sinh(x)) doesn't returns all solutions
``` python
In [2]: solveset(sinh(x))
Out[2]: {0}
```

### LINKED ISSUE #9606: `solveset_complex(sinh(x), x)` returns `FiniteSet(0)`
```
>>> x = Symbol('x')
>>> solveset_complex(sinh(x), x)
FiniteSet(0)
```

It should return `ImageSet(Lambda(_n, _n*I*pi), Integers())`

### PR DESCRIPTION
**Pull request by smichr**
closes #10722 
closes #10552 
fixes #9531
fixes #9606

---
- [ ] fix auto-rewrite problems

### CODE REVIEW

**Comment by kshitij10496** [comment]
I was trying to fix the same issue but I think your implementation is better for handling trigonometric and hyperbolic equations.

However, I would like to point that we need better `ImageSet` handler for unions and intersections.
For example:
```python
In []: solveset(sinh(x).rewrite(exp),x)
Out[]: {n⋅ⅈ⋅π | n ∊ ℤ}

In []: solveset(sinh(x),x)
Out[]: ({2⋅n⋅ⅈ⋅π | n ∊ ℤ} ∩ ℂ) ∪ (ℂ ∩ {ⅈ⋅(2⋅n⋅π + π) | n ∊ ℤ})
```
Both the outputs represent the same mathematical value.
And this should be reflected by simplifying the second result.

If its fine by you, I will work on it.

**Comment by smichr** [comment]
> If its fine by you, I will work on it.

Yes, I've seen this as well. If you have some ideas, go ahead and give it a go. Don't forget about `factor_terms`:  `factor_terms(2*n*pi + pi) -> pi*(2*n + 1)`.

**Comment by kshitij10496** [comment]
@smichr Thanks for the advice.
This surely will come in handy.

File: sympy/solvers/solveset.py [hunk 1]
```diff
@@ -366,26 +366,33 @@ def _solve_as_rational(f, symbol, domain):
...
+            cov = exp(symbol)
+        g, h = [i.subs(cov, y) for i in (G, H)]
+        if g.has(symbol) or h.has(symbol):
+            continue
+        # the change of variables succeeded
+        soln = solveset_complex(g, y) - solveset_complex(h, y)
+        if isinstance(soln, FiniteSet):
+            result = Union(*[
+                invert_complex(cov, s, symbol)[1] for s in soln])
+            result = Intersection(result, domain)
```
Comment (smichr) [inline sympy/solvers/solveset.py]: is it necessary to use the domain here or will that be taken care of by solveset?

File: sympy/solvers/tests/test_solveset.py [hunk 2]
```diff
@@ -567,15 +568,27 @@ def test_solve_abs():
...
+    n = Dummy('n')
+    pi_2n = 2*n*pi
+    l1 = log(sqrt(-2 + sqrt(5)))
+
+    assert solveset_real(sinh(x) + sech(x), x) == FiniteSet(l1)
+
+    l2 = log(sqrt(2 + sqrt(5)))
+    Int = lambda x: ImageSet(Lambda(n, x), S.Integers)
+    assert solveset(sinh(x) + sech(x), x) == Union(
+        imageset(n,I*pi_2n + l1, S.Integers),
```
Comment (smichr) [inline sympy/solvers/tests/test_solveset.py]: XXX writing these args as `Int(I*pi_2n + l1)`, etc... did not work. Why?

**Comment by smichr** [comment]
stale

**Pull request closed without merging**